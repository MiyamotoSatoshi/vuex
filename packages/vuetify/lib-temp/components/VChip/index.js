import VChip from './VChip';
export { VChip };
export default VChip;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi9zcmMvY29tcG9uZW50cy9WQ2hpcC9pbmRleC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxPQUFPLEtBQUssTUFBTSxTQUFTLENBQUE7QUFFM0IsT0FBTyxFQUFFLEtBQUssRUFBRSxDQUFBO0FBQ2hCLGVBQWUsS0FBSyxDQUFBIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IFZDaGlwIGZyb20gJy4vVkNoaXAnXG5cbmV4cG9ydCB7IFZDaGlwIH1cbmV4cG9ydCBkZWZhdWx0IFZDaGlwXG4iXX0=