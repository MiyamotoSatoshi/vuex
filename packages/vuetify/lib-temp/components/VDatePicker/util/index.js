import createNativeLocaleFormatter from './createNativeLocaleFormatter';
import monthChange from './monthChange';
import pad from './pad';
export { createNativeLocaleFormatter, monthChange, pad };
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi8uLi9zcmMvY29tcG9uZW50cy9WRGF0ZVBpY2tlci91dGlsL2luZGV4LmpzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLE9BQU8sMkJBQTJCLE1BQU0sK0JBQStCLENBQUE7QUFDdkUsT0FBTyxXQUFXLE1BQU0sZUFBZSxDQUFBO0FBQ3ZDLE9BQU8sR0FBRyxNQUFNLE9BQU8sQ0FBQTtBQUV2QixPQUFPLEVBQ0wsMkJBQTJCLEVBQzNCLFdBQVcsRUFDWCxHQUFHLEVBQ0osQ0FBQSIsInNvdXJjZXNDb250ZW50IjpbImltcG9ydCBjcmVhdGVOYXRpdmVMb2NhbGVGb3JtYXR0ZXIgZnJvbSAnLi9jcmVhdGVOYXRpdmVMb2NhbGVGb3JtYXR0ZXInXG5pbXBvcnQgbW9udGhDaGFuZ2UgZnJvbSAnLi9tb250aENoYW5nZSdcbmltcG9ydCBwYWQgZnJvbSAnLi9wYWQnXG5cbmV4cG9ydCB7XG4gIGNyZWF0ZU5hdGl2ZUxvY2FsZUZvcm1hdHRlcixcbiAgbW9udGhDaGFuZ2UsXG4gIHBhZFxufVxuIl19