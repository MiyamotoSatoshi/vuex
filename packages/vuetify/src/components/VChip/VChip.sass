// Imports
@import './_variables.scss'

// Theme
.v-chip:not(.v-chip--outlined)
  &.primary,
  &.secondary,
  &.accent,
  &.success,
  &.error,
  &.warning,
  &.info
    color: map-deep-get($material-dark, 'text', 'primary')

+theme(v-chip) using ($material)
  border-color: map-get($material, 'dividers')
  color: map-deep-get($material, 'text', 'primary')

  &:not(.v-chip--active)
    background: map-get($material, 'chips')

  +states($material)

// Block
.v-chip
  align-items: center
  cursor: default
  display: inline-flex
  line-height: 20px
  max-width: 100%
  outline: none
  overflow: hidden
  padding: 0 12px
  position: relative
  text-decoration: none
  transition-duration: $chip-transition-duration
  transition-property: box-shadow, opacity
  transition-timing-function: $chip-transition-fn
  vertical-align: middle
  white-space: nowrap

  &:before
    background-color: currentColor
    bottom: 0
    border-radius: inherit
    content: ''
    left: 0
    opacity: 0
    position: absolute
    pointer-events: none
    right: 0
    top: 0

  // Important is needed to account for new avatar structure
  .v-avatar
    height: $chip-avatar-size !important
    min-width: $chip-avatar-size !important
    width: $chip-avatar-size !important

  .v-icon
    font-size: $chip-icon-size

  .v-avatar,
  .v-icon
    &--left
      margin-left: $chip-icon-margin-before
      margin-right: $chip-icon-margin-after

    &--right
      margin-left: $chip-icon-margin-after
      margin-right: $chip-icon-margin-before

    +rtl()
      &--left
        margin-left: $chip-icon-margin-after
        margin-right: $chip-icon-margin-before

      &--right
        margin-left: $chip-icon-margin-before
        margin-right: $chip-icon-margin-after

  &:not(.v-chip--no-color)
    .v-icon
      color: inherit

// Elements
.v-chip__close.v-icon
  font-size: $chip-close-size
  max-height: $chip-close-size
  max-width: $chip-close-size
  user-select: none

  &.v-icon--right
    margin-right: $chip-icon-right-margin-after

    +rtl()
      margin-left: $chip-icon-right-margin-after
      margin-right: $chip-icon-right-margin-before

  &:hover,
  &:focus,
  &:active
    opacity: .72

.v-chip__content
  align-items: center
  display: inline-flex
  height: 100%
  max-width: 100%

// Modifiers
.v-chip--active
  .v-icon
    color: inherit

.v-chip--link
  &::before
    transition: opacity $primary-transition

  &:focus::before
    opacity: $chip-link-focus-opacity

.v-chip--clickable
  cursor: pointer

  &:active
    +elevation(2)

.v-chip--disabled
  opacity: .4
  pointer-events: none
  user-select: none

.v-chip__filter
  // TODO: Figure out why the absence
  // of this breaks the transition
  max-width: 24px

  // Increase specificity to
  // overwrite theme style
  &.v-icon
    color: inherit

  &.expand-x-transition-leave-active,
  &.expand-x-transition-enter
    margin: 0

.v-chip--pill
  .v-chip__filter
    margin-right: $chip-pill-filter-margin

  .v-avatar
    height: $chip-pill-avatar-size !important
    width: $chip-pill-avatar-size !important

  .v-avatar
    &--left
      margin-left: $chip-pill-avatar-margin-before

    &--right
      margin-right: $chip-pill-avatar-margin-before

    +rtl()
      &--left
        margin-left: chip-pill-avatar-margin-after
        margin-right: $chip-pill-avatar-margin-before

      &--right
        margin-left: $chip-pill-avatar-margin-before
        margin-right: chip-pill-avatar-margin-after

.v-chip--label
  border-radius: $chip-label-border-radius !important

.v-chip.v-chip--outlined
  border-width: $icon-outlined-border-width
  border-style: solid

  &:not(.v-chip--active):before
    opacity: 0

  &.v-chip--active:before
    opacity: 0.08

  .v-icon
    color: inherit

  // Needs increased specificity
  &.v-chip.v-chip
    background-color: transparent !important

// Needs increased specificity
// to overwrite theme color
.v-chip.v-chip--selected
  background: transparent

  &:after
    opacity: .28

.v-chip
  @each $name, $size in $icon-sizes
    &.v-size--#{$name}
      border-radius: #{map-get($size, 'height') / 2}px
      font-size: #{map-get($size, 'font-size')}px
      height: #{map-get($size, 'height')}px
