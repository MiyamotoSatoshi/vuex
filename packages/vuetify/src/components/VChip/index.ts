export { default as VChip } from './VChip'
