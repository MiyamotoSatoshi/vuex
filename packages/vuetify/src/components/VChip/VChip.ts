// Styles
import './VChip.sass'

// Types
import { VNode } from 'vue'
import mixins from '../../util/mixins'

// Components
import { VExpandXTransition } from '../transitions'
import VIcon from '../VIcon'

// Mixins
import Colorable from '../../mixins/colorable'
import { factory as GroupableFactory } from '../../mixins/groupable'
import Themeable from '../../mixins/themeable'
import { factory as ToggleableFactory } from '../../mixins/toggleable'
import Sizeable from '../../mixins/sizeable'

// Directives
import Ripple from '../../directives/ripple'

// Utilities
import { deprecate } from '../../util/console'

// Types
import { PropValidator } from 'vue/types/options'

/* @vue/component */
export default mixins(
  Colorable,
  GroupableFactory('chipGroup'),
  Sizeable,
  Themeable,
  ToggleableFactory('inputValue')
).extend({
  name: 'v-chip',

  directives: { Ripple },

  props: {
    close: Boolean,
    closeIcon: {
      type: String,
      default: '$vuetify.icons.delete'
    },
    disabled: Boolean,
    draggable: Boolean,
    filter: Boolean,
    filterIcon: {
      type: String,
      default: '$vuetify.icons.complete'
    },
    label: Boolean,
    link: Boolean,
    outline: Boolean,
    outlined: Boolean,
    pill: Boolean,
    ripple: {
      type: [Boolean, Object],
      default: null
    },
    // Used for selects/tagging
    selected: Boolean,
    textColor: String,
    value: null as any as PropValidator<any>
  },

  computed: {
    classes (): object {
      return {
        'v-chip--clickable': this.isClickable,
        'v-chip--disabled': this.disabled,
        'v-chip--draggable': this.draggable,
        'v-chip--label': this.label,
        'v-chip--no-color': !this.color,
        'v-chip--outlined': this.hasOutline,
        'v-chip--pill': this.pill,
        'v-chip--removable': this.hasClose,
        'v-chip--selected': !this.disabled && this.isActive,
        ...this.themeClasses,
        ...this.sizeableClasses,
        ...this.groupClasses
      }
    },
    hasClose (): boolean {
      return Boolean(
        this.close ||
        this.$listeners['click:close']
      )
    },
    hasOutline (): boolean {
      return this.outline || this.outlined
    },
    isClickable (): boolean {
      return Boolean(
        this.link ||
        this.chipGroup ||
        this.$listeners.click ||
        this.$listeners['!click'] ||
        this.$attrs.tabindex
      )
    }
  },

  methods: {
    genFilter (): VNode {
      const children = []

      if (this.isActive) {
        children.push(
          this.$createElement(VIcon, {
            staticClass: 'v-chip__filter',
            props: { left: true }
          }, this.filterIcon)
        )
      }

      return this.$createElement(VExpandXTransition, children)
    },
    genClose (): VNode {
      return this.$createElement(VIcon, {
        staticClass: 'v-chip__close',
        props: {
          right: true
        },
        on: {
          click: (e: Event) => {
            e.stopPropagation()

            this.$emit('click:close')
          }
        }
      }, this.closeIcon)
    },
    genContent (): VNode {
      return this.$createElement('span', {
        staticClass: 'v-chip__content'
      }, [
        this.filter && this.genFilter(),
        this.$slots.default,
        this.hasClose && this.genClose()
      ])
    }
  },

  created () {
    if (this.outline) deprecate('outline', 'outlined', this)
    if (this.selected) deprecate('selected', 'value', this)
  },

  render (h): VNode {
    const data = this.setBackgroundColor(this.color, {
      staticClass: 'v-chip',
      class: this.classes,
      attrs: {
        ...this.$attrs,
        draggable: this.draggable ? 'true' : undefined,
        tabindex: this.disabled ? -1 : this.$attrs.tabindex || +this.isClickable - 1
      },
      directives: [{
        name: 'ripple',
        value: this.ripple != null ? this.ripple : !this.disabled
      }],
      on: this.$listeners
    })

    const color = this.textColor || (this.hasOutline && this.color)

    if (!this.disabled && this.chipGroup) {
      this._g(data, { click: this.toggle })
    }

    return h(
      'span',
      this.setTextColor(color, data),
      [this.genContent()]
    )
  }
})
