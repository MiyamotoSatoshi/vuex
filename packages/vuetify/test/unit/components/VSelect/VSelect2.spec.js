import { test } from '@/test'
import { keyCodes } from '@/util/helpers'
import VSelect from '@/components/VSelect/VSelect'
import VChip from '@/components/VChip'

test('VSelect2', ({ mount, compileToFunctions }) => {
  const app = document.createElement('div')
  app.setAttribute('data-app', true)
  document.body.appendChild(app)

  // Inspired by https://github.com/vuetifyjs/vuetify/pull/1425 - Thanks @kevmo314
  it('should open the select when focused and enter, space, up or down are pressed', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: ['foo', 'bar']
      }
    })

    const input = wrapper.first('input')

    for (const key of ['up', 'down', 'space', 'enter']) {
      input.trigger('focus')
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMenuActive).toBe(false)
      input.trigger(`keydown.${key}`)
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMenuActive).toBe(true)
      wrapper.vm.isMenuActive = false // reset for next iteration
      await wrapper.vm.$nextTick()
    }
  })

  it('should not open the select when readonly and focused and enter, space, up or down are pressed', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: ['foo', 'bar'],
        readonly: true
      }
    })

    const input = wrapper.first('input')

    for (const key of ['up', 'down', 'space', 'enter']) {
      input.trigger('focus')
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMenuActive).toBe(false)
      input.trigger(`keydown.${key}`)
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMenuActive).toBe(false)
      await wrapper.vm.$nextTick()
    }
  })

  it('should clear input value', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        clearable: true,
        items: ['foo', 'bar'],
        value: 'foo'
      }
    })

    const clear = wrapper.find('.v-icon')[0]

    const input = jest.fn()
    wrapper.vm.$on('input', input)

    await wrapper.vm.$nextTick()
    expect(wrapper.vm.internalValue).toBe('foo')
    clear.trigger('click')

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.internalValue).toBe(undefined)
    expect(input).toHaveBeenCalledWith(undefined)
  })

  it('should be clearable with prop, dirty and single select', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        clearable: true,
        items: [1, 2],
        value: 1
      }
    })

    const clear = wrapper.first('.v-icon')

    await wrapper.vm.$nextTick()
    expect(wrapper.vm.internalValue).toBe(1)
    expect(wrapper.html()).toMatchSnapshot()

    clear.trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.internalValue).toBe(undefined)
    expect(wrapper.vm.isMenuActive).toBe(false)
  })

  it('should be clearable with prop, dirty and multi select', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        clearable: true,
        items: [1, 2],
        multiple: true,
        value: [1]
      }
    })

    const clear = wrapper.first('.v-icon')

    const change = jest.fn()
    wrapper.vm.$on('change', change)

    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    clear.trigger('click')
    await wrapper.vm.$nextTick()
    expect(change).toHaveBeenCalledWith([])
    expect(wrapper.vm.isMenuActive).toBe(false)
  })

  it('should prepopulate selectedItems', () => {
    const items = ['foo', 'bar', 'baz']

    const wrapper = mount(VSelect, {
      propsData: {
        items,
        value: 'foo'
      }
    })

    const wrapper2 = mount(VSelect, {
      propsData: {
        items,
        multiple: true,
        value: ['foo', 'bar']
      }
    })

    const wrapper3 = mount(VSelect, {
      propsData: {
        items,
        value: null
      }
    })

    expect(wrapper.vm.selectedItems).toHaveLength(1)
    expect(wrapper2.vm.selectedItems).toHaveLength(2)
    expect(wrapper3.vm.selectedItems).toHaveLength(0)
  })

  it('should show input with placeholder and not dirty', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        placeholder: 'foo'
      }
    })

    expect(wrapper.find('input')[0].hasStyle('display', 'block'))
  })

  it('should not show input with placeholder and dirty', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: ['bar'],
        placeholder: 'foo',
        value: 'bar'
      }
    })

    expect(wrapper.find('input')[0].hasStyle('display', 'none'))
  })

  // #1704
  it('should populate select when using value as an object', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: [
          { text: 'foo', value: { id: 1 } },
          { text: 'foo', value: { id: 2 } }
        ],
        multiple: true,
        value: [{ id: 1 }]
      }
    })

    await wrapper.vm.$nextTick()

    const selections = wrapper.find('.v-select__selection')

    expect(selections.length).toBeGreaterThan(0)
  })

  // Discovered when working on #1704
  it('should remove item when re-selecting it', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: [
          { text: 'bar', value: { id: 1 } },
          { text: 'foo', value: { id: 2 } }
        ],
        multiple: true,
        value: [{ id: 1 }]
      }
    })

    expect(wrapper.vm.selectedItems).toHaveLength(1)
    wrapper.trigger('click')
    const item = wrapper.first('div.v-list__tile__action')
    item.trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.selectedItems).toHaveLength(0)
  })

  it('should open menu when cleared with open-on-clear', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        clearable: true,
        openOnClear: true,
        items: [1],
        value: 1
      }
    })

    const clear = wrapper.first('.v-input__icon--clear .v-icon')

    clear.trigger('click')

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isMenuActive).toBe(true)

  })

  it('should react to different key down', async () => {
    const wrapper = mount(VSelect)
    const blur = jest.fn()
    wrapper.vm.$on('blur', blur)

    wrapper.vm.onKeyDown({ keyCode: keyCodes.tab })

    expect(blur).toBeCalled()
    expect(wrapper.vm.isMenuActive).toBe(false)

    for (let keyCode of [keyCodes.enter, keyCodes.space, keyCodes.up, keyCodes.down]) {
      wrapper.vm.onKeyDown({ keyCode })
      expect(wrapper.vm.isMenuActive).toBe(true)

      wrapper.vm.isMenuActive = false
      expect(wrapper.vm.isMenuActive).toBe(false)
    }
  })

  it('should select an item !multiple', async () => {
    const wrapper = mount(VSelect)

    const input = jest.fn()
    const change = jest.fn()
    wrapper.vm.$on('input', input)
    wrapper.vm.$on('change', change)

    wrapper.vm.selectItem('foo')

    expect(wrapper.vm.internalValue).toBe('foo')
    expect(input).toBeCalledWith('foo')
    expect(input).toHaveBeenCalledTimes(1)

    await wrapper.vm.$nextTick()

    expect(change).toBeCalledWith('foo')
    expect(change).toHaveBeenCalledTimes(1)

    wrapper.setProps({ returnObject: true })

    const item = { foo: 'bar' }
    wrapper.vm.selectItem(item)

    expect(wrapper.vm.internalValue).toBe(item)
    expect(input).toBeCalledWith(item)
    expect(input).toHaveBeenCalledTimes(2)

    await wrapper.vm.$nextTick()

    expect(change).toBeCalledWith(item)
    expect(change).toHaveBeenCalledTimes(2)
  })

  it('should disable v-list-tile', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        items: [{ text: 'foo', disabled: true, id: 0 }]
      }
    })

    const selectItem = jest.fn()
    wrapper.setMethods({ selectItem })

    let el = wrapper.first('.v-list__tile')

    el.element.click()

    expect(selectItem).not.toBeCalled()

    wrapper.setProps({
      items: [{ text: 'foo', disabled: false, id: 0 }]
    })

    await wrapper.vm.$nextTick()

    el.element.click()

    expect(selectItem).toBeCalled()
  })

  it('should update menu status and focus when menu closes', async () => {
    const wrapper = mount(VSelect)
    const menu = wrapper.vm.$refs.menu

    wrapper.setData({
      isMenuActive: true,
      isFocused: true
    })

    expect(wrapper.vm.isMenuActive).toBe(true)
    expect(wrapper.vm.isFocused).toBe(true)

    await wrapper.vm.$nextTick()

    expect(menu.isActive).toBe(true)

    menu.isActive = false

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isMenuActive).toBe(false)
    expect(wrapper.vm.isFocused).toBe(false)
  })

  it('should use specified value', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        items: ['foo']
      }
    })
  })

  it('should update model when chips are removed', async () => {
    const selectItem = jest.fn()
    const wrapper = mount(VSelect, {
      propsData: {
        chips: true,
        deletableChips: true,
        items: ['foo'],
        value: 'foo'
      },
      methods: { selectItem }
    })

    const input = jest.fn()
    const change = jest.fn()

    wrapper.vm.$on('input', input)

    expect(wrapper.vm.internalValue).toEqual('foo')
    wrapper.first('.v-chip__close').trigger('click')

    expect(input).toHaveBeenCalledTimes(1)

    wrapper.setProps({
      items: ['foo', 'bar'],
      multiple: true,
      value: ['foo', 'bar']
    })
    wrapper.vm.$on('change', change)
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.internalValue).toEqual(['foo', 'bar'])
    wrapper.first('.v-chip__close').trigger('click')

    await wrapper.vm.$nextTick()

    expect(selectItem).toHaveBeenCalledTimes(1)
  })

  it('should set selected index', async () => {
    const onFocus = jest.fn()
    const wrapper = mount(VSelect, {
      propsData: {
        chips: true,
        deletableChips: true,
        multiple: true,
        items: ['foo', 'bar', 'fizz', 'buzz'],
        value: ['foo', 'bar', 'fizz', 'buzz']
      },
      methods: { onFocus }
    })

    expect(wrapper.vm.selectedIndex).toBe(-1)

    const foo = wrapper.first('.v-chip--select-multi')
    foo.trigger('click')

    expect(wrapper.vm.selectedIndex).toBe(0)
    expect(onFocus).toHaveBeenCalledTimes(1)

    wrapper.find('.v-chip--select-multi')[1].trigger('click')

    expect(wrapper.vm.selectedIndex).toBe(1)
    expect(onFocus).toHaveBeenCalledTimes(2)

    wrapper.setProps({ disabled: true })

    await wrapper.vm.$nextTick()

    wrapper.first('.v-chip--select-multi').trigger('click')

    expect(wrapper.vm.selectedIndex).toBe(1)
    expect(onFocus).toHaveBeenCalledTimes(2)
  })

  it('should not duplicate items after items update when caching is turned on', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        cacheItems: true,
        returnObject: true,
        itemText: 'text',
        itemValue: 'id',
        items: [],
      }
    })

    wrapper.setProps({ items: [{ id: 1, text: 'A' }] })
    expect(wrapper.vm.computedItems).toHaveLength(1)
    wrapper.setProps({ items: [{ id: 1, text: 'A' }] })
    expect(wrapper.vm.computedItems).toHaveLength(1)
  })

  it('should cache items', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        cacheItems: true,
        items: []
      }
    })

    wrapper.setProps({ items: ['bar', 'baz'] })
    expect(wrapper.vm.computedItems).toHaveLength(2)

    wrapper.setProps({ items: ['foo'] })
    expect(wrapper.vm.computedItems).toHaveLength(3)

    wrapper.setProps({ items: ['bar'] })
    expect(wrapper.vm.computedItems).toHaveLength(3)
  })

  it('should cache items passed via prop', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        cacheItems: true,
        items: [1, 2, 3, 4]
      }
    })

    expect(wrapper.vm.computedItems).toHaveLength(4)

    wrapper.setProps({ items: [5] })
    expect(wrapper.vm.computedItems).toHaveLength(5)
  })

  it('should have an affix', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        prefix: '$',
        suffix: 'lbs'
      }
    })

    expect(wrapper.first('.v-text-field__prefix').element.innerHTML).toBe('$')
    expect(wrapper.first('.v-text-field__suffix').element.innerHTML).toBe('lbs')

    wrapper.setProps({ prefix: undefined, suffix: undefined })

    await wrapper.vm.$nextTick()

    expect(wrapper.find('.v-text-field__prefix').length).toBe(0)
    expect(wrapper.find('.v-text-field__suffix').length).toBe(0)
  })

  it('should use custom clear icon cb', () => {
    const clearIconCb = jest.fn()
    const wrapper = mount(VSelect, {
      propsData: {
        clearable: true,
        items: ['foo'],
        value: 'foo'
      }
    })

    wrapper.vm.$on('click:clear', clearIconCb)
    wrapper.first('.v-input__icon--clear .v-icon').trigger('click')

    expect(clearIconCb).toBeCalled()
  })

  it('should populate select[multiple=false] when using value as an object', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: [
          { text: 'foo', value: { id: { subid: 1 } } },
          { text: 'foo', value: { id: { subid: 2 } } }
        ],
        multiple: false,
        value: { id: { subid: 2 } }
      }
    })

    await wrapper.vm.$nextTick()

    const selections = wrapper.find('.v-select__selection')

    expect(selections.length).toEqual(1)
  })

  it('should add color to selected index', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        multiple: true,
        items: ['foo', 'bar'],
        value: ['foo']
      }
    })

    wrapper.vm.selectedIndex = 0

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not react to click when disabled', async () => {
    const wrapper = mount(VSelect, {
      propsData: { items: ['foo', 'bar'] }
    })
    const slot = wrapper.first('.v-input__slot')
    const input = wrapper.first('input')

    expect(wrapper.vm.isMenuActive).toBe(false)
    slot.trigger('click')
    expect(wrapper.vm.isMenuActive).toBe(true)

    wrapper.setData({ isMenuActive: false })
    wrapper.setProps({ disabled: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.isMenuActive).toBe(false)

    slot.trigger('click')
    expect(wrapper.vm.isMenuActive).toBe(false)
  })

  it('should set the menu index', () => {
    const wrapper = mount(VSelect)

    expect(wrapper.vm.getMenuIndex()).toBe(-1)

    wrapper.vm.setMenuIndex(1)

    expect(wrapper.vm.getMenuIndex()).toBe(1)
  })
})
