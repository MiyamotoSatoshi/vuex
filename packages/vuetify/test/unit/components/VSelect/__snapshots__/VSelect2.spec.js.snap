// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSelect2 should add color to selected index 1`] = `

<div role="combobox"
     class="v-input v-text-field v-select v-input--is-label-active v-input--is-dirty theme--light"
>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-select__slot">
        <div class="v-select__selections">
          <div class="v-select__selection v-select__selection--comma primary--text">
            foo
          </div>
          <input readonly="readonly"
                 type="text"
                 autocomplete="on"
                 aria-readonly="false"
          >
        </div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--append">
            <i aria-hidden="true"
               class="v-icon material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div class="v-menu">
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <div class="v-messages__wrapper">
        </div>
      </div>
    </div>
  </div>
</div>

`;

exports[`VSelect2 should be clearable with prop, dirty and multi select 1`] = `

<div role="combobox"
     class="v-input v-text-field v-select v-input--is-label-active v-input--is-dirty theme--light"
>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-select__slot">
        <div class="v-select__selections">
          <div class="v-select__selection v-select__selection--comma">
            1
          </div>
          <input readonly="readonly"
                 type="text"
                 autocomplete="on"
                 aria-readonly="false"
          >
        </div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--clear">
            <i aria-hidden="true"
               class="v-icon v-icon--link material-icons theme--light"
            >
              clear
            </i>
          </div>
        </div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--append">
            <i aria-hidden="true"
               class="v-icon material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div class="v-menu">
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <div class="v-messages__wrapper">
        </div>
      </div>
    </div>
  </div>
</div>

`;

exports[`VSelect2 should be clearable with prop, dirty and single select 1`] = `

<div role="combobox"
     class="v-input v-text-field v-select v-input--is-label-active v-input--is-dirty theme--light"
>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-select__slot">
        <div class="v-select__selections">
          <div class="v-select__selection v-select__selection--comma">
            1
          </div>
          <input readonly="readonly"
                 type="text"
                 autocomplete="on"
                 aria-readonly="false"
          >
        </div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--clear">
            <i aria-hidden="true"
               class="v-icon v-icon--link material-icons theme--light"
            >
              clear
            </i>
          </div>
        </div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--append">
            <i aria-hidden="true"
               class="v-icon material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div class="v-menu">
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <div class="v-messages__wrapper">
        </div>
      </div>
    </div>
  </div>
</div>

`;
